pkg/driver/aws_profile.go

```diff
@@ -0,0 +1,124 @@
+package driver
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	awsProfileName                = "s3-csi"
+	awsProfileConfigFilename      = "s3-csi-config"
+	awsProfileCredentialsFilename = "s3-csi-credentials"
+	awsProfileFilePerm            = fs.FileMode(0400) // only owner readable
+)
+
+// An AWSProfile represents an AWS profile with it's credentials and config files.
+type AWSProfile struct {
+	Name            string
+	ConfigPath      string
+	CredentialsPath string
+}
+
+// CreateAWSProfile creates an AWS Profile with credentials and config files from given credentials.
+// Created credentials and config files can be clean up with `CleanupAWSProfile`.
+func CreateAWSProfile(basepath string, accessKeyID string, secretAccessKey string, sessionToken string) (AWSProfile, error) {
+	name := awsProfileName
+
+	configPath := filepath.Join(basepath, awsProfileConfigFilename)
+	err := writeAWSProfileFile(configPath, configFileContents(name))
+	if err != nil {
+		return AWSProfile{}, fmt.Errorf("aws-profile: Failed to create config file %s: %v", configPath, err)
+	}
+
+	credentialsPath := filepath.Join(basepath, awsProfileCredentialsFilename)
+	err = writeAWSProfileFile(credentialsPath, credentialsFileContents(name, accessKeyID, secretAccessKey, sessionToken))
+	if err != nil {
+		return AWSProfile{}, fmt.Errorf("aws-profile: Failed to create credentials file %s: %v", credentialsPath, err)
+	}
+
+	return AWSProfile{
+		Name:            name,
+		ConfigPath:      configPath,
+		CredentialsPath: credentialsPath,
+	}, nil
+}
+
+// CleanupAWSProfile cleans up credentials and config files created in given `basepath` via `CreateAWSProfile`.
+func CleanupAWSProfile(basepath string) error {
+	configPath := filepath.Join(basepath, awsProfileConfigFilename)
+	if err := os.Remove(configPath); err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return fmt.Errorf("aws-profile: Failed to remove config file %s: %v", configPath, err)
+		}
+	}
+
+	credentialsPath := filepath.Join(basepath, awsProfileCredentialsFilename)
+	if err := os.Remove(credentialsPath); err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return fmt.Errorf("aws-profile: Failed to credentials config file %s: %v", credentialsPath, err)
+		}
+	}
+
+	return nil
+}
+
+func writeAWSProfileFile(path string, content string) error {
+	err := os.WriteFile(path, []byte(content), awsProfileFilePerm)
```

Comment:
Not sure I follow, we already pass the 3rd argument for file mode, but the problem is:
>  If the file does not exist, WriteFile creates it with permissions perm (before umask); otherwise WriteFile truncates it before writing, __without changing permissions__.

Reply:
That's only if the file doesn't exist. We expect this file to not exist at the time of writing, right? Would it be ok to delete and overwrite it if it does exist?

Reply:
The issue with the current version is that there's a short period where the file has the default permissions associated with umask rather than what we want.